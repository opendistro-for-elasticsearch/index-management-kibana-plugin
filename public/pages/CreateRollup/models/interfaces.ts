/*
 * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import { ManagedCatIndex } from "../../../../server/models/interfaces";
import { FieldItem } from "../../../../models/interfaces";

export interface MappingsItem {
  properties?: {};
  type?: string;
}

export interface DimensionItem {
  sequence: number;
  field: FieldItem;
  aggregationMethod: string;
  interval?: number;
}

export interface MetricItem {
  source_field: FieldItem;
  all: boolean;
  min: boolean;
  max: boolean;
  sum: boolean;
  avg: boolean;
  value_count: boolean;
}

//From backend
export interface RollupDimensionItem {
  date_histogram?: {
    fixed_interval?: string;
    calendar_interval?: string;
    source_field: string;
    target_field: string;
    timezone: string;
  };
  terms?: {
    source_field: string;
    target_field: string;
  };
  histogram?: {
    source_field: string;
    target_field: string;
    interval: number;
  };
}

export interface RollupMetricItem {
  source_field: string;
  metrics: [
    {
      min?: Object;
      max?: Object;
      sum?: Object;
      avg?: Object;
      value_count?: Object;
    }
  ];
}
