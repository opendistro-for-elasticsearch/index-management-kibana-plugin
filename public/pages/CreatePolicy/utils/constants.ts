/*
 * Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

export const DEFAULT_POLICY = JSON.stringify(
  {
    policy: {
      description: "A simple default policy that changes the replica count between hot and cold states.",
      default_state: "hot",
      states: [
        {
          name: "hot",
          actions: [{ replica_count: { number_of_replicas: 5 } }],
          transitions: [{ state_name: "cold", conditions: { min_index_age: "30d" } }],
        },
        {
          name: "cold",
          actions: [{ replica_count: { number_of_replicas: 2 } }],
          transitions: [],
        },
      ],
    },
  },
  null,
  4
);
