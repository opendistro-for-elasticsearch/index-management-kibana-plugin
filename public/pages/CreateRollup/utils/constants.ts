/*
 * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

//Added last_updated_time and start_time to avoid error, but they should be system generated.
//Added a delay as 0, but this should be optional, so null should be fine.
export const EMPTY_ROLLUP = JSON.stringify({
  rollup: {
    continuous: false,
    description: "",
    dimensions: [
      {
        date_histogram: {
          source_field: "",
          fixed_interval: "2ms",
          timezone: "Africa/Abidjan",
        },
      },
    ],
    enabled: false,
    metrics: [],
    page_size: 1000,
    roles: [],
    schedule: {
      interval: {
        start_time: 234802,
        period: 1,
        unit: "MINUTES",
      },
    },
    source_index: "",
    target_index: "",
  },
});

export const FixedTimeunitOptions = [
  { value: "ms", text: "Millisecond(s)" },
  { value: "s", text: "Second(s)" },
  { value: "m", text: "Minute(s)" },
  { value: "h", text: "Hour(s)" },
  { value: "d", text: "Day(s)" },
];

export const DelayTimeunitOptions = [
  { value: "SECONDS", text: "Second(s)" },
  { value: "MINUTES", text: "Minute(s)" },
  { value: "HOURS", text: "Hour(s)" },
  { value: "DAYS", text: "Day(s)" },
];

export const CalendarTimeunitOptions = [
  { value: "MINUTES", text: "Minute(s)" },
  { value: "HOURS", text: "Hour(s)" },
  { value: "DAYS", text: "Day(s)" },
];

export const AddFieldsColumns = [
  {
    field: "label",
    name: "Field name",
    sortable: true,
  },
  {
    field: "type",
    name: "Field type",
    sortable: true,
    render: (type: string | undefined) => (type == null || type == undefined ? "-" : type),
  },
];
