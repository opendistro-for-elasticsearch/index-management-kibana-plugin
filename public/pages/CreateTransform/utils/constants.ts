/*
 * Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

export const EMPTY_TRANSFORM = JSON.stringify({
  transform: {
    description: "",
    enabled: true,
    page_size: 1000,
    roles: [],
    schedule: {
      interval: {
        start_time: 234802,
        period: 1,
        unit: "MINUTES",
      },
    },
    source_index: "",
    target_index: "",
  },
});

export const FixedTimeunitOptions = [
  { value: "ms", text: "Millisecond(s)" },
  { value: "s", text: "Second(s)" },
  { value: "m", text: "Minute(s)" },
  { value: "h", text: "Hour(s)" },
  { value: "d", text: "Day(s)" },
];

export const DelayTimeunitOptions = [
  { value: "SECONDS", text: "Second(s)" },
  { value: "MINUTES", text: "Minute(s)" },
  { value: "HOURS", text: "Hour(s)" },
  { value: "DAYS", text: "Day(s)" },
];

export const CalendarTimeunitOptions = [
  { value: "m", text: "Minute" },
  { value: "h", text: "Hour" },
  { value: "d", text: "Day" },
  { value: "w", text: "Week" },
  { value: "M", text: "Month" },
  { value: "q", text: "Quarter" },
  { value: "y", text: "Year" },
];

export const ScheduleIntervalTimeunitOptions = [
  { value: "MINUTES", text: "Minute(s)" },
  { value: "HOURS", text: "Hour(s)" },
  { value: "DAYS", text: "Day(s)" },
];

export const AddFieldsColumns = [
  {
    field: "label",
    name: "Field name",
    sortable: true,
  },
  {
    field: "type",
    name: "Field type",
    sortable: true,
    render: (type: string | undefined) => (type == null || type == undefined ? "-" : type),
  },
];

export const ExecutionFrequencyDefinitionOptions = [
  { value: "fixed", text: "Define by fixed interval" },
  { value: "cron", text: "Define by cron expression" },
];

export const DefaultSampleDataSize = 50;
