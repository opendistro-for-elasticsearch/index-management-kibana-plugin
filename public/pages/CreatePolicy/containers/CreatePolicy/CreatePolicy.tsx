/*
 * Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import React, { ChangeEvent, Component, Fragment } from "react";
import { EuiSpacer, EuiTitle, EuiFlexGroup, EuiFlexItem, EuiButton, EuiButtonEmpty, EuiCallOut, EuiLink, EuiIcon } from "@elastic/eui";
import chrome from "ui/chrome";
import { toastNotifications } from "ui/notify";
import queryString from "query-string";
import { RouteComponentProps } from "react-router";
import { DEFAULT_POLICY } from "../../utils/constants";
import DefinePolicy from "../../components/DefinePolicy";
import ConfigurePolicy from "../../components/ConfigurePolicy";
import { Policy } from "../../../../../models/interfaces";
import { PolicyService } from "../../../../services";
import { BREADCRUMBS, DOCUMENTATION_URL, ROUTES } from "../../../../utils/constants";
import { getErrorMessage } from "../../../../utils/helpers";

interface CreatePolicyProps extends RouteComponentProps {
  isEdit: boolean;
  policyService: PolicyService;
}

interface CreatePolicyState {
  policyId: string;
  policyIdError: string;
  jsonString: string;
  policySeqNo: number | null;
  policyPrimaryTerm: number | null;
  submitError: string;
  isSubmitting: boolean;
  hasSubmitted: boolean;
}

export default class CreatePolicy extends Component<CreatePolicyProps, CreatePolicyState> {
  _isMounted: boolean = false;

  constructor(props: CreatePolicyProps) {
    super(props);

    this.state = {
      policySeqNo: null,
      policyPrimaryTerm: null,
      policyId: "",
      policyIdError: "",
      submitError: "",
      jsonString: "",
      isSubmitting: false,
      hasSubmitted: false,
    };
  }

  componentDidMount = async (): Promise<void> => {
    this._isMounted = true;
    chrome.breadcrumbs.set([BREADCRUMBS.INDEX_MANAGEMENT, BREADCRUMBS.INDEX_POLICIES]);
    if (this.props.isEdit) {
      const { id } = queryString.parse(this.props.location.search);
      if (typeof id === "string" && !!id) {
        chrome.breadcrumbs.push(BREADCRUMBS.EDIT_POLICY);
        chrome.breadcrumbs.push({ text: id });
        await this.getPolicyToEdit(id);
      } else {
        toastNotifications.addDanger(`Invalid policy id: ${id}`);
        this.props.history.push(ROUTES.INDEX_POLICIES);
      }
    } else {
      chrome.breadcrumbs.push(BREADCRUMBS.CREATE_POLICY);
      this.setState({ jsonString: DEFAULT_POLICY });
    }
  };

  componentWillUnmount(): void {
    this._isMounted = false;
  }

  getPolicyToEdit = async (policyId: string): Promise<void> => {
    try {
      const { policyService } = this.props;
      const response = await policyService.getPolicy(policyId);
      if (response.ok) {
        this.setState({
          policySeqNo: response.response.seqNo,
          policyPrimaryTerm: response.response.primaryTerm,
          policyId: response.response.id,
          jsonString: JSON.stringify({ policy: response.response.policy }, null, 4),
        });
      } else {
        toastNotifications.addDanger(`Could not load the policy: ${response.error}`);
        this.props.history.push(ROUTES.INDEX_POLICIES);
      }
    } catch (err) {
      toastNotifications.addDanger(getErrorMessage(err, "Could not load the policy"));
      this.props.history.push(ROUTES.INDEX_POLICIES);
    }
  };

  onCreate = async (policyId: string, policy: Policy): Promise<void> => {
    const { policyService } = this.props;
    try {
      const response = await policyService.putPolicy(policy, policyId);
      if (response.ok) {
        toastNotifications.addSuccess(`Created policy: ${response.response._id}`);
        this.props.history.push(ROUTES.INDEX_POLICIES);
      } else {
        this.setState({ submitError: response.error });
      }
    } catch (err) {
      this.setState({ submitError: getErrorMessage(err, "There was a problem creating the policy") });
    }
  };

  onUpdate = async (policyId: string, policy: Policy): Promise<void> => {
    try {
      const { policyService } = this.props;
      const { policyPrimaryTerm, policySeqNo } = this.state;
      if (policySeqNo == null || policyPrimaryTerm == null) {
        toastNotifications.addDanger("Could not update policy without seqNo and primaryTerm");
        return;
      }
      const response = await policyService.putPolicy(policy, policyId, policySeqNo, policyPrimaryTerm);
      if (response.ok) {
        toastNotifications.addSuccess(`Updated policy: ${response.response._id}`);
        this.props.history.push(ROUTES.INDEX_POLICIES);
      } else {
        this.setState({ submitError: response.error });
      }
    } catch (err) {
      this.setState({ submitError: getErrorMessage(err, "There was a problem updating the policy") });
    }
  };

  onCancel = (): void => {
    if (this.props.isEdit) this.props.history.goBack();
    else this.props.history.push(ROUTES.INDEX_POLICIES);
  };

  onChange = (e: ChangeEvent<HTMLInputElement>): void => {
    const { hasSubmitted } = this.state;
    const policyId = e.target.value;
    if (hasSubmitted) this.setState({ policyId, policyIdError: policyId ? "" : "Required" });
    else this.setState({ policyId });
  };

  onChangeJSON = (value: string): void => {
    this.setState({ jsonString: value });
  };

  onAutoIndent = (): void => {
    try {
      const parsedJSON = JSON.parse(this.state.jsonString);
      this.setState({ jsonString: JSON.stringify(parsedJSON, null, 4) });
    } catch (err) {
      // do nothing
    }
  };

  onSubmit = async (): Promise<void> => {
    const { isEdit } = this.props;
    const { policyId, jsonString } = this.state;
    this.setState({ submitError: "", isSubmitting: true, hasSubmitted: true });
    try {
      if (!policyId) {
        this.setState({ policyIdError: "Required" });
      } else {
        const policy = JSON.parse(jsonString);
        if (isEdit) await this.onUpdate(policyId, policy);
        else await this.onCreate(policyId, policy);
      }
    } catch (err) {
      toastNotifications.addDanger("Invalid Policy JSON");
      console.error(err);
    }

    if (this._isMounted) this.setState({ isSubmitting: false });
  };

  renderEditCallOut = (): React.ReactNode | null => {
    const { isEdit } = this.props;
    if (!isEdit) return null;

    return (
      <Fragment>
        <EuiCallOut
          title="Edits to the policy are not automatically applied to indices that are already being managed by this policy."
          iconType="questionInCircle"
        >
          <p>
            This ensures that any update to a policy doesn't harm indices that are running under an older version of the policy. To carry
            over your edits to these indices, please use the "Change Policy" under "Managed Indices" to reapply the policy after submitting
            your edits.{" "}
            <EuiLink href={DOCUMENTATION_URL} target="_blank">
              Learn more <EuiIcon type="popout" size="s" />
            </EuiLink>
          </p>
        </EuiCallOut>
        <EuiSpacer />
      </Fragment>
    );
  };

  render() {
    const { isEdit } = this.props;
    const { policyId, policyIdError, jsonString, submitError, isSubmitting } = this.state;

    let hasJSONError = false;
    try {
      JSON.parse(jsonString);
    } catch (err) {
      hasJSONError = true;
    }

    return (
      <div style={{ padding: "25px 50px" }}>
        <EuiTitle size="l">
          <h1>{isEdit ? "Edit" : "Create"} policy</h1>
        </EuiTitle>
        <EuiSpacer />
        {this.renderEditCallOut()}
        <ConfigurePolicy policyId={policyId} policyIdError={policyIdError} isEdit={isEdit} onChange={this.onChange} />
        <EuiSpacer />
        <DefinePolicy jsonString={jsonString} onChange={this.onChangeJSON} onAutoIndent={this.onAutoIndent} hasJSONError={hasJSONError} />
        <EuiSpacer />
        {submitError && (
          <EuiCallOut title="Sorry, there was an error" color="danger" iconType="alert">
            <p>{submitError}</p>
          </EuiCallOut>
        )}
        <EuiSpacer />
        <EuiFlexGroup alignItems="center" justifyContent="flexEnd">
          <EuiFlexItem grow={false}>
            <EuiButtonEmpty onClick={this.onCancel} data-test-subj="createPolicyCancelButton">
              Cancel
            </EuiButtonEmpty>
          </EuiFlexItem>
          <EuiFlexItem grow={false}>
            <EuiButton fill onClick={this.onSubmit} isLoading={isSubmitting} data-test-subj="createPolicyCreateButton">
              {isEdit ? "Update" : "Create"}
            </EuiButton>
          </EuiFlexItem>
        </EuiFlexGroup>
      </div>
    );
  }
}
