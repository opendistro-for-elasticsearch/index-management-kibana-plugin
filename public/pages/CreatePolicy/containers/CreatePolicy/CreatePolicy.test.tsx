/*
 * Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import React from "react";
import { render, wait, fireEvent } from "@testing-library/react";
import CreatePolicy from "./CreatePolicy";
import { ServicesConsumer, ServicesContext } from "../../../../services";
import { MemoryRouter as Router, Redirect, Route, RouteComponentProps, Switch } from "react-router";
import { browserServicesMock } from "../../../../../test/mocks";
import { BrowserServices } from "../../../../models/interfaces";
import { ModalProvider, ModalRoot } from "../../../../components/Modal";
import { DEFAULT_POLICY } from "../../utils/constants";
import userEvent from "@testing-library/user-event";
import { toastNotifications } from "ui/notify";
import { ROUTES } from "../../../../utils/constants";

jest.mock("../../components/DefinePolicy", () => require("../../components/DefinePolicy/__mocks__/DefinePolicyMock"));

function renderCreatePolicyWithRouter(initialEntries = ["/"]) {
  return {
    ...render(
      <Router initialEntries={initialEntries}>
        <ServicesContext.Provider value={browserServicesMock}>
          <ServicesConsumer>
            {(services: BrowserServices | null) =>
              services && (
                <ModalProvider>
                  <ModalRoot services={services} />
                  <Switch>
                    <Route
                      path={ROUTES.CREATE_POLICY}
                      render={(props: RouteComponentProps) => (
                        <CreatePolicy {...props} isEdit={false} policyService={services.policyService} />
                      )}
                    />
                    <Route
                      path={ROUTES.EDIT_POLICY}
                      render={(props: RouteComponentProps) => (
                        <CreatePolicy {...props} isEdit={true} policyService={services.policyService} />
                      )}
                    />
                    <Route path={ROUTES.INDEX_POLICIES} render={(props: RouteComponentProps) => <div>Testing Policies</div>} />
                    <Redirect from="/" to={ROUTES.CREATE_POLICY} />
                  </Switch>
                </ModalProvider>
              )
            }
          </ServicesConsumer>
        </ServicesContext.Provider>
      </Router>
    ),
  };
}

describe("<CreatePolicy /> spec", () => {
  it("renders the create component", () => {
    const { container } = renderCreatePolicyWithRouter();

    expect(container.firstChild).toMatchSnapshot();
  });

  it("renders the edit component", async () => {
    browserServicesMock.policyService.getPolicy = jest
      .fn()
      .mockResolvedValue({ ok: true, response: { seqNo: 1, primaryTerm: 5, id: "some_policy", policy: JSON.parse(DEFAULT_POLICY) } });
    const { container } = renderCreatePolicyWithRouter([`${ROUTES.EDIT_POLICY}?id=some_policy`]);

    await wait();

    expect(container.firstChild).toMatchSnapshot();
  });

  it("routes back to policies if given bad id", async () => {
    const { getByText } = renderCreatePolicyWithRouter([`${ROUTES.EDIT_POLICY}?id=one&id=two`]);

    await wait(() => getByText("Testing Policies"));
    expect(toastNotifications.addDanger).toHaveBeenCalledTimes(1);
    expect(toastNotifications.addDanger).toHaveBeenCalledWith("Invalid policy id: one,two");
  });

  it("routes back to policies if getPolicy gracefully fails", async () => {
    browserServicesMock.policyService.getPolicy = jest.fn().mockResolvedValue({ ok: false, error: "some error" });
    const { getByText } = renderCreatePolicyWithRouter([`${ROUTES.EDIT_POLICY}?id=some_id`]);

    await wait(() => getByText("Testing Policies"));
    expect(toastNotifications.addDanger).toHaveBeenCalledTimes(1);
    expect(toastNotifications.addDanger).toHaveBeenCalledWith("Could not load the policy: some error");
  });

  it("routes back to policies if getPolicy gracefully fails", async () => {
    browserServicesMock.policyService.getPolicy = jest.fn().mockRejectedValue(new Error("another error"));
    const { getByText } = renderCreatePolicyWithRouter([`${ROUTES.EDIT_POLICY}?id=some_id`]);

    await wait(() => getByText("Testing Policies"));
    expect(toastNotifications.addDanger).toHaveBeenCalledTimes(1);
    expect(toastNotifications.addDanger).toHaveBeenCalledWith("another error");
  });

  it("disallows editing policy ID when in edit", async () => {
    browserServicesMock.policyService.getPolicy = jest
      .fn()
      .mockResolvedValue({ ok: true, response: { seqNo: 1, primaryTerm: 5, id: "some_id", policy: JSON.parse(DEFAULT_POLICY) } });
    const { getByDisplayValue, getByPlaceholderText } = renderCreatePolicyWithRouter([`${ROUTES.EDIT_POLICY}?id=some_id`]);

    await wait(() => getByDisplayValue("some_id"));

    expect(getByPlaceholderText("hot_cold_workflow")).toHaveAttribute("readonly");
  });

  it("shows error for policyId input when clicking create", async () => {
    const { getByTestId, queryByText, getByPlaceholderText } = renderCreatePolicyWithRouter();

    expect(queryByText("Required")).toBeNull();

    userEvent.click(getByTestId("createPolicyCreateButton"));

    expect(queryByText("Required")).not.toBeNull();

    fireEvent.focus(getByPlaceholderText("hot_cold_workflow"));
    userEvent.type(getByPlaceholderText("hot_cold_workflow"), `some_policy_id`);
    fireEvent.blur(getByPlaceholderText("hot_cold_workflow"));

    expect(queryByText("Required")).toBeNull();
  });

  it("routes you back to policies and shows a success toaster when successfully creating a policy", async () => {
    browserServicesMock.policyService.putPolicy = jest.fn().mockResolvedValue({ ok: true, response: { _id: "some_policy_id" } });
    const { getByText, getByTestId, getByPlaceholderText } = renderCreatePolicyWithRouter();

    fireEvent.focus(getByPlaceholderText("hot_cold_workflow"));
    userEvent.type(getByPlaceholderText("hot_cold_workflow"), `some_policy_id`);
    fireEvent.blur(getByPlaceholderText("hot_cold_workflow"));

    userEvent.click(getByTestId("createPolicyCreateButton"));

    await wait(() => getByText("Testing Policies"));
    expect(toastNotifications.addSuccess).toHaveBeenCalledWith("Created policy: some_policy_id");
  });

  it("shows a danger toaster when getting graceful error from create policy", async () => {
    browserServicesMock.policyService.putPolicy = jest.fn().mockResolvedValue({ ok: false, error: "bad policy" });
    const { getByText, getByTestId, getByPlaceholderText } = renderCreatePolicyWithRouter();

    fireEvent.focus(getByPlaceholderText("hot_cold_workflow"));
    userEvent.type(getByPlaceholderText("hot_cold_workflow"), `some_policy_id`);
    fireEvent.blur(getByPlaceholderText("hot_cold_workflow"));

    userEvent.click(getByTestId("createPolicyCreateButton"));

    await wait(() => getByText("bad policy"));
  });

  it("shows a danger toaster when getting error from create policy", async () => {
    browserServicesMock.policyService.putPolicy = jest.fn().mockRejectedValue(new Error("this is an error"));
    const { getByText, getByTestId, getByPlaceholderText } = renderCreatePolicyWithRouter();

    fireEvent.focus(getByPlaceholderText("hot_cold_workflow"));
    userEvent.type(getByPlaceholderText("hot_cold_workflow"), `some_policy_id`);
    fireEvent.blur(getByPlaceholderText("hot_cold_workflow"));

    userEvent.click(getByTestId("createPolicyCreateButton"));

    await wait(() => getByText("this is an error"));
  });

  it("routes you back to policies and shows a success toaster when successfully updating a policy", async () => {
    browserServicesMock.policyService.putPolicy = jest.fn().mockResolvedValue({ ok: true, response: { _id: "some_policy_id" } });
    browserServicesMock.policyService.getPolicy = jest
      .fn()
      .mockResolvedValue({ ok: true, response: { seqNo: 1, primaryTerm: 5, id: "some_policy_id", policy: JSON.parse(DEFAULT_POLICY) } });
    const { getByText, getByTestId, getByDisplayValue } = renderCreatePolicyWithRouter([`${ROUTES.EDIT_POLICY}?id=some_policy_id`]);

    await wait(() => getByDisplayValue("some_policy_id"));

    userEvent.click(getByTestId("createPolicyCreateButton"));

    await wait(() => getByText("Testing Policies"));
    expect(toastNotifications.addSuccess).toHaveBeenCalledWith("Updated policy: some_policy_id");
  });

  it("shows error when getting graceful error from create policy", async () => {
    browserServicesMock.policyService.putPolicy = jest.fn().mockResolvedValue({ ok: false, error: "bad policy" });
    browserServicesMock.policyService.getPolicy = jest
      .fn()
      .mockResolvedValue({ ok: true, response: { seqNo: 1, primaryTerm: 5, id: "some_policy_id", policy: JSON.parse(DEFAULT_POLICY) } });
    const { getByTestId, getByDisplayValue, getByText } = renderCreatePolicyWithRouter([`${ROUTES.EDIT_POLICY}?id=some_policy_id`]);

    await wait(() => getByDisplayValue("some_policy_id"));

    userEvent.click(getByTestId("createPolicyCreateButton"));

    await wait(() => getByText("bad policy"));
  });

  it("shows a danger toaster when getting error from create policy", async () => {
    browserServicesMock.policyService.putPolicy = jest.fn().mockRejectedValue(new Error("this is an error"));
    browserServicesMock.policyService.getPolicy = jest
      .fn()
      .mockResolvedValue({ ok: true, response: { seqNo: 1, primaryTerm: 5, id: "some_policy_id", policy: JSON.parse(DEFAULT_POLICY) } });
    const { getByText, getByTestId, getByDisplayValue } = renderCreatePolicyWithRouter([`${ROUTES.EDIT_POLICY}?id=some_policy_id`]);

    await wait(() => getByDisplayValue("some_policy_id"));

    userEvent.click(getByTestId("createPolicyCreateButton"));

    await wait(() => getByText("this is an error"));
  });

  it("brings you back to policies when clicking cancel", async () => {
    const { getByTestId, getByText } = renderCreatePolicyWithRouter();

    userEvent.click(getByTestId("createPolicyCancelButton"));

    await wait(() => getByText("Testing Policies"));
  });
});
