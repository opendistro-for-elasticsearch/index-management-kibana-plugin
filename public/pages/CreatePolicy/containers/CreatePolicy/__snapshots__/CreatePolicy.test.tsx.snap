// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CreatePolicy /> spec renders the create component 1`] = `
<div
  style="padding: 25px 50px;"
>
  <h1
    class="euiTitle euiTitle--large"
  >
    Create
     policy
  </h1>
  <div
    class="euiSpacer euiSpacer--l"
  />
  <div
    class="euiPanel euiPanel--paddingMedium"
    style="padding-left: 0px; padding-right: 0px;"
  >
    <div
      class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
      style="padding: 0px 10px;"
    >
      <div
        class="euiFlexItem"
      >
        <h3
          class="euiTitle euiTitle--small"
        >
          Name policy
        </h3>
      </div>
    </div>
    <hr
      class="euiHorizontalRule euiHorizontalRule--full euiHorizontalRule--marginXSmall"
    />
    <div
      style="padding: initial;"
    >
      <div
        style="padding-left: 10px;"
      >
        <div
          class="euiText euiText--extraSmall"
        >
          <p>
            Policies let you automatically perform administrative operations on indices.
          </p>
        </div>
        <div
          class="euiSpacer euiSpacer--s"
        />
        <div
          class="euiFormRow"
          id="some_html_id-row"
        >
          <div
            class="euiFormRow__labelWrapper"
          >
            <label
              aria-invalid="false"
              class="euiFormLabel euiFormRow__label"
              for="some_html_id"
            >
              Policy ID
            </label>
          </div>
          <div
            class="euiFormRow__fieldWrapper"
          >
            <div
              class="euiFormControlLayout"
            >
              <div
                class="euiFormControlLayout__childrenWrapper"
              >
                <input
                  aria-describedby="some_html_id-help"
                  class="euiFieldText"
                  id="some_html_id"
                  placeholder="hot_cold_workflow"
                  type="text"
                  value=""
                />
              </div>
            </div>
            <div
              class="euiFormHelpText euiFormRow__text"
              id="some_html_id-help"
            >
              Specify a unique ID that is easy to recognize and remember.
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="euiSpacer euiSpacer--l"
  />
  <div
    class="euiPanel euiPanel--paddingMedium"
    style="padding-left: 0px; padding-right: 0px;"
  >
    <div
      class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
      style="padding: 0px 10px;"
    >
      <div
        class="euiFlexItem"
      >
        <h3
          class="euiTitle euiTitle--small"
        >
          Define policy
        </h3>
      </div>
      <div
        class="euiFlexItem euiFlexItem--flexGrowZero"
      >
        <div
          class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
        >
          <div
            class="euiFlexItem"
          >
            <span
              class="euiToolTipAnchor"
            >
              <button
                class="euiButton euiButton--primary"
                type="button"
              >
                <span
                  class="euiButtonContent euiButton__content"
                >
                  EuiIconMock
                  <span
                    class="euiButton__text"
                  >
                    Copy
                  </span>
                </span>
              </button>
            </span>
          </div>
          <div
            class="euiFlexItem"
          >
            <button
              class="euiButton euiButton--primary euiButton-isDisabled"
              disabled=""
              type="button"
            >
              <span
                class="euiButtonContent euiButton__content"
              >
                EuiIconMock
                <span
                  class="euiButton__text"
                >
                  Auto Indent
                </span>
              </span>
            </button>
          </div>
        </div>
      </div>
    </div>
    <hr
      class="euiHorizontalRule euiHorizontalRule--full euiHorizontalRule--marginXSmall"
    />
    <div
      style="padding: initial;"
    >
      <div
        style="padding-left: 10px;"
      >
        <div
          class="euiText euiText--extraSmall"
        >
          <p>
            Create a policy with a JSON configuration file. This can be added directly in the code editor below.
          </p>
        </div>
      </div>
      <div
        class="euiSpacer euiSpacer--s"
      />
      <textarea
        aria-label="Code Editor"
        class="euiTextArea euiTextArea--resizeVertical"
        rows="6"
      />
    </div>
  </div>
  <div
    class="euiSpacer euiSpacer--l"
  />
  
  <div
    class="euiSpacer euiSpacer--l"
  />
  <div
    class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentFlexEnd euiFlexGroup--directionRow euiFlexGroup--responsive"
  >
    <div
      class="euiFlexItem euiFlexItem--flexGrowZero"
    >
      <button
        class="euiButtonEmpty euiButtonEmpty--primary"
        data-test-subj="createPolicyCancelButton"
        type="button"
      >
        <span
          class="euiButtonContent euiButtonEmpty__content"
        >
          <span
            class="euiButtonEmpty__text"
          >
            Cancel
          </span>
        </span>
      </button>
    </div>
    <div
      class="euiFlexItem euiFlexItem--flexGrowZero"
    >
      <button
        class="euiButton euiButton--primary euiButton--fill"
        data-test-subj="createPolicyCreateButton"
        type="button"
      >
        <span
          class="euiButtonContent euiButton__content"
        >
          <span
            class="euiButton__text"
          >
            Create
          </span>
        </span>
      </button>
    </div>
  </div>
</div>
`;

exports[`<CreatePolicy /> spec renders the edit component 1`] = `
<div>
  Testing Policies
</div>
`;

exports[`<CreatePolicy /> spec routes back to policies if given bad id 1`] = `null`;
